litestar/controller/generic.py

```diff
@@ -188,20 +206,106 @@ def path_param_type(self) -> str:
+        if self.update_dto is not Empty:
+            return cast("type[AbstractDTO[ModelT]]", self.update_dto)
+
+        dto_type = self._dto_type[self.model_type]  # type: ignore[index]
+
+        class _UpdateDTO(dto_type):  # type: ignore[misc,valid-type]
+            config = DTOConfig(partial=True)
+
+        return _UpdateDTO
+
+    @cached_property
+    def _return_dto(self) -> type[AbstractDTO[ModelT]] | None:
+        if self.return_dto is not Empty:
+            return cast("type[AbstractDTO[ModelT]] | None", self.return_dto)
+
+        app = cast("Litestar", self.ownership_layers[0])
+        if any(
+            plugin.supports_type(FieldDefinition.from_annotation(self.model_type))
+            for plugin in app.plugins.serialization
+        ):
+            # since there is a serialization plugin in place, a return dto type is not required.
+            return None
+
+        return cast("type[AbstractDTO[ModelT]]", self._dto_type[self.model_type])  # type: ignore[index]
+
+    @cached_property
+    def _dto_type(self) -> type[AbstractDTO]:
+        if self.dto_type is not Empty:
+            return cast("type[AbstractDTO]", self.dto_type)
+        if is_dataclass(self.model_type):
+            return DataclassDTO
+        if is_struct_class(self.model_type):
+            return MsgspecDTO
+        if is_pydantic_model_class(self.model_type):
+            from litestar.contrib.pydantic import PydanticDTO
+
+            return PydanticDTO
+        if is_piccolo_class(self.model_type):
+            from litestar.contrib.piccolo import PiccoloDTO
+
+            return PiccoloDTO
+        if is_sqlalchemy_model(self.model_type):
+            from litestar.contrib.sqlalchemy import SQLAlchemyDTO
+
+            return SQLAlchemyDTO
+
+        if filtered_data_dtos := [
+            layer.dto for layer in self.ownership_layers if layer.dto is not Empty and layer.dto is not None
+        ]:
+            return cast("type[AbstractDTO]", filtered_data_dtos[-1])
+
+        raise ImproperlyConfiguredException(
+            f"cannot determine DTO type for {self.model_type.__name__} on {type(self).__name__} "
+            f"Generic Controller. Please specify a `dto_type` to have dtos automatically generated, or "
+            f"specify the `create_dto`, `update_dto` and `return_dto` attributes to avoid dto generation.`"
+        )
+
     def create_repository(
         self, *, request: Request[Any, Any, Any], **kwargs: Any
     ) -> AbstractAsyncRepository[ModelT] | AbstractSyncRepository[ModelT]:
         kwargs["request"] = request
         return self.repository_type(**kwargs)
 
-    async def create_instance(self, data: ModelT, request: Request[Any, Any, Any]) -> ModelT:
-        result = self.create_repository(request=request).add(data=data)
+    async def create_instance(self, data: DTOData[ModelT], request: Request[Any, Any, Any]) -> ModelT:
+        result = self.create_repository(request=request).add(data=data.as_builtins())
```

Comment:
I must be missing something here, why not `DTOData.create_instance()` That to me looks like it will save you from all the edits on the repositories.

Reply:
instance is missing an ID or is partial (depending on the method), if its pydantic etc. it will fail validation

Reply:
I see what is happening now.

I would suggest an alternative API that would require sightly more input from the user (but leads to less code for the library):

1. Have a required input to the generic controller that is an "id_factory". The results of calling the factory will be passed to `create_instance` hence no validation errors.
2. This one is more tricky, we need to obtain the whole model from the repository and pass it to `update_instance`. This can be done without requiring any input from the user, we already have the `repository` and the `id` (from the url path).

What do you think?

Reply:
You are welcome to make a parallel PR with these, let's see what's what and decide between the patterns.